### ISSUE TITLE
Consider enabling - (dash/hypen) in prerelease strings

### ISSUE THREAD

**Issue by aharpervc** [rubygems/rubygems#3086]
I'm having a problem and would like to a feature.

My current problem is:

I would like to publish prerelease versions with dashes, eg something like `1.2.0.pre-update-logo`. This appears to not be possible as of rubygems version 3.0.4:

```log
irb(main):001:0> Gem::Version.new('1.2.0.pre-update-logo')
=> #<Gem::Version "1.2.0.pre.pre.update.pre.logo">
```

The `-`'s are all replaced by `.pre.` which is fairly unhelpful (except possibly the first replacement). That happens here: https://github.com/rubygems/rubygems/blob/d8fa21e82b5669880331e2fe236b2cede1d6c4fb/lib/rubygems/version.rb#L221

Dashes in prerelease versions are allowed by the SemVer spec: 

> A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version. Identifiers MUST comprise only ASCII alphanumerics and hyphen [0-9A-Za-z-]

Note that I'm not asking about "FULL SEMVER", but only about dashes in prerelease labels

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [x] Publishing a library
  - [ ] The command line `gem`
  - [x] Other

Other issues:

- Support for full Semantic versioning #592
- Match rubygems not-quite-Semver versioning for pre-release versioning #86

Here are my current environment details:

```
$ gem env version
RubyGems Environment:
  - RUBYGEMS VERSION: 3.0.4
  - RUBY VERSION: 2.5.3 (2018-10-18 patchlevel 105) [x86_64-linux]
(etc, n/a)
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by esotericpig** [comment]
I had a similar thing with bundler.

My Gemspec:

```ruby
spec.version = '0.4.9-alpha'
```

According to Semantic Version v2, this is correct.

I then created a release and tag on GitHub as `v0.4.9-alpha`.

Then I was surprised when I did this command:

```
$ bundle exec rake release
```

as it then creates the release and tag:

```
v0.4.9.pre.alpha
```

I decided to just remove `-alpha` because then I'd have to do all of my releases manually.

I think it'd be good to change this for `bundler/gem_tasks` (which is `gem_helper`).

You can see my [example releases](https://github.com/esotericpig/UncleKryon-server/releases) here.

**Comment by pboling** [comment]
Just ran into this, and now I'm not sure what the correct way to version my gem is.  I'm trying to follow SemVer, but RubyGems doesn't allow it.

I'd also like for my version string to match the tagged and released version, as well as the version people put in their gemfiles.

**Comment by martinemde** [comment]
I think the fundamental challenge here is that rubygems get composed into dash separated strings from the gem name, version, and platform. This means that `require_bench-1.0.4-alpha2` would have an identical filename to the gem `require_bench` version `1.0.4` for platform `alpha2`. As far as I understand, this is why the versions dashes are converted.

I'm not commenting on how hard this is to solve, or whether or not it will get solved, but it is not trivial.

**Comment by indirect** [comment]
As you all have noticed, you can in fact denote prerelease versions by putting a dash in the version number.

RubyGems is not able to put that dash into the filename (as described by Martin above), so the dash is always converted to `.pre.` instead. For example, this Gemfile installs the gem whose file is `conjur-debify-3.0.3.pre.1933.gem`:

```ruby
source "https://rubygems.org"
gem "conjur-debify", "3.0.3-1933"
```

That said, if you want your version string, tag, .gem filename, etc, to all be identical, you need to use `.pre.` instead of `-`.

In conclusion, RubyGems _does_ allow a dash in prerelease version strings, but we cannot put that dash into the final .gem filename without any modification.

**Comment by aharpervc** [comment]
> In conclusion, RubyGems _does_ allow a dash in prerelease version strings

What parameter value should I pass to `Gem::Version.new` to have the resulting value show dashes, per the example above?

**Comment by martinemde** [comment]
Whoa! I used the github mobile app to quote reply and it edited your comment! Very sorry. Let me see if I can revert it. How strange.

**Comment by aharpervc** [comment]
> If you specify a gem with version 1.2.3-alpha

How can this version be defined? Also, how would you do `1.2.3-alpha-with-dashes`? When I run that in irb, I get this, which is the same as what I originally posted above and remains unresolved:

```
irb(main):002:0> Gem::Version.new('1.2.3-alpha-with-dashes')
=> Gem::Version.new("1.2.3.pre.alpha.pre.with.pre.dashes")
```

**Comment by martinemde** [comment]
If you name your gem and add it to bundler in the same way, it should work regardless of how the file and version are transformed internally. I agree it's not pretty but SemVer didn't exist when rubygems adopted this pattern.

**Comment by pboling** [comment]
When pre-release versions are involved `Gem::Version.new` comparison operator doesn't sort them properly.
```
gem_pkgs = File.join("pkg", "*.gem")
gems = Dir[gem_pkgs]
# Should be newest last!
gems.sort_by! { |gem| Gem::Version.new(gem[VERSION_REGEX]) }
gem_pkg = gems.last
puts gem_pkg # => "pkg/my_gem-1.0.4.pre.alpha.7.gem"
```
It should have been "pkg/my_gem-1.0.4.gem"
![Screenshot 2023-10-10 at 12 02 18](https://github.com/rubygems/rubygems/assets/19505/fbf64345-ff72-4927-ab6d-b479c4384336)

**Comment by martinemde** [comment]
I don't get the same result:

```ruby
irb(main):008> [Gem::Version.new("1.0.4-alpha1"), Gem::Version.new("1.0.4-alpha12"), Gem::Version.new("1.0.4")].sort
=> [Gem::Version.new("1.0.4.pre.alpha1"), Gem::Version.new("1.0.4.pre.alpha12"), Gem::Version.new("1.0.4")]
```

**Comment by pboling** [comment]
That was exactly it.  Thanks!  I assumed erroneously that it would handle .gem in some special way (because it worked until I used it with pre-release versions).
